### DISCUSSION TITLE
`Signer` API discussion

### DISCUSSION THREAD

**Discussion by frol** [near/near-api-rs#36] [Ideas]
*I moved this discussion from our [Telegram group](https://t.me/NEAR_Tools_Community_Group/4044)*

Original feedback (from me):

> `Signer::new(Signer::from_seed_phrase())` looks really odd to me. `T::from_*` should return `T`, but it is not the case somehow

@akorchyn responded:

> Yeah, agree. I want to have it simple. And to have ability to get any implementation at one place. Thast's why I put in the wrapper these constructors of other types. But it's confusing.
> Maybe, enum. `Signer::new(SignerType::Seed)`, `SignerType::SeedWithHdPhrase`, etc

**Comment by frol** [comment]
@akorchyn Do you really need `Signer` **struct** at all? I would expect that the method where you pass (`with_signer()`), you'd just expect anything that implements `Signer` **trait** (and KeystoreSigner, LedgerSigner, and SignersPool to implement Signer **trait**)

**Reply by akorchyn** [reply]
Well, we also need to cache nonce for consecutive calls. I think it's easier to maintain by having simple wrapper than implement it for every signer

**Reply by frol** [reply]
I see... Let me think about this challenge for a bit.
